### PR TITLE
Added multiple pointer actions for select and hold

### PR DESCRIPTION
**Pull request by StephenHodgson**
# Overview

Added the ability to specify multiple select and hold actions as well as single axis actions and not just binary ones.

Added a select and hold threshold value for single axis press actions.

## Breaking Changes

- Any prefabs with scripts that derive from `BaseControllerPointer` will need to have their hold and select actions updated.image

### CODE REVIEW

**Comment by SimonDarksideJ** [comment]
Looks good to me, nice and clean.  But would like @provencher to also cast his eye over.

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs [hunk 1]
```diff
@@ -47,12 +47,23 @@ public abstract class BaseControllerPointer : ControllerPoseSynchronizer,
...
         [SerializeField]
-        [Tooltip("The hold action that will enable the raise the input event for this pointer.")]
-        private MixedRealityInputAction activeHoldAction = MixedRealityInputAction.None;
+        [Tooltip("The hold actions that will enable the raise the input event for this pointer.")]
+        private MixedRealityInputAction[] holdActions = { MixedRealityInputAction.None };
+
+        [SerializeField]
+        [Range(0.01f, 1f)]
+        [Tooltip("The threshold before raising a single axis hold action.")]
+        private float holdThreshold = 0.25f;
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: Maybe `holdPressThreshold`?

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs [hunk 2]
```diff
@@ -47,12 +47,23 @@ public abstract class BaseControllerPointer : ControllerPoseSynchronizer,
...
+        [FormerlySerializedAs("pointerActions")]
+        [Tooltip("The actions that will enable the raise the input event for this pointer.")]
+        private MixedRealityInputAction[] selectActions = { MixedRealityInputAction.None };
 
         [SerializeField]
-        [Tooltip("The action that will enable the raise the input event for this pointer.")]
-        private MixedRealityInputAction pointerAction = MixedRealityInputAction.None;
+        [Range(0.01f, 1f)]
+        [Tooltip("The threshold before raising a single axis select action.")]
+        private float selectThreshold = 0.25f;
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: Maybe `selectPressThreshold`?

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs [hunk 3]
```diff
@@ -47,12 +47,23 @@ public abstract class BaseControllerPointer : ControllerPoseSynchronizer,
...
+        [Tooltip("The hold actions that will enable the raise the input event for this pointer.")]
+        private MixedRealityInputAction[] holdActions = { MixedRealityInputAction.None };
+
+        [SerializeField]
+        [Range(0.01f, 1f)]
+        [Tooltip("The threshold before raising a single axis hold action.")]
+        private float holdThreshold = 0.25f;
+
+        [SerializeField]
+        [FormerlySerializedAs("pointerActions")]
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: This line can get removed.

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs (hunk 1 above)
Comment (provencher) [reply XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: I set it to 0.4 for my work framework

Comment (provencher) [reply XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: Is this just a deadzone value?

**Comment by provencher** [comment]
I don't really understand the reason for this PR - It feels a bit backwards.

Reading this code, it seems like the input system will trigger a pointer down call on this pointer, which then triggers a select call back to the input system. This change then adds more possible input system events to catch when the pointer down call is triggered, which feels backwards.

IMO, the pointer shouldn't be triggering input system calls. It should receive events from the input system, which are then fed down into the pointer interaction receiver. A pointer is something that should exist alongside Controllers, who should bear the responsibility of triggering the input system.

**Comment by StephenHodgson** [comment]
>IMO, the pointer shouldn't be triggering input system calls. It should receive events from the input system, which are then fed down into the pointer interaction receiver. A pointer is something that should exist alongside Controllers, who should bear the responsibility of triggering the input system.

Agreed but that's some crap left over from old mrtk (thanks to Kurtis).

As much as I'd love to change that today that's beyond the scope of this pr.

**Comment by SimonDarksideJ** [comment]
Agreed, this is a good first step along a journey to get to the position we want, without radicalising the framework.

@provencher could you create an issue proposal with the remaining gaps you see, so we can plan the next steps?

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs (hunk 1 above)
Comment (StephenHodgson) [reply XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Pointers/BaseControllerPointer.cs]: It's the value threshold that determines if the input down/up event is fired by the single axis action data. (I suppose it is a deadzone? 🤷‍♂️)

**Comment by StephenHodgson** [comment]
I wanna test this before merging. I saw some issues with the pointer results when testing on ML

**Comment by SimonDarksideJ** [comment]
Just changed my review to avoid accidental merging :D. Also put a call out in General for others to acquaint themselves with the change

**Comment by SimonDarksideJ** [comment]
You have several notable actions against yourself in this PR @StephenHodgson , can you either do the updates or remove the comments / actions?

**Pull request closed without merging**